### DISCUSSION TITLE
Use PostCss configuration from Next.js for SSR with css-in-js

### DISCUSSION THREAD

**Discussion by ties-v** [vercel/next.js#12271] [Help] [answered]
Hi there,

At this moment I work on a project that is using a JSS based component library. To enable server side rendered styles, I followed the [material-ui example](https://github.com/mui-org/material-ui/tree/master/examples/nextjs) which works fine.

To improve performance, I want to minify and autoprefix the extracted css when server side rendering. The material-ui docs website does this by incorporating postcss in `_document.js`  (_[source](https://github.com/mui-org/material-ui/blob/24cdb5bbcb2015fadeea2de80266c66a7c2a3858/docs/pages/_document.js#L16-L27)_):
```js
let prefixer;
let cleanCSS;
if (process.env.NODE_ENV === 'production') {
  /* eslint-disable global-require */
  const postcss = require('postcss');
  const autoprefixer = require('autoprefixer');
  const CleanCSS = require('clean-css');
  /* eslint-enable global-require */

  prefixer = postcss([autoprefixer]);
  cleanCSS = new CleanCSS();
}
```
And then minify it after collecting the css in getInitialProps (_[source](https://github.com/mui-org/material-ui/blob/24cdb5bbcb2015fadeea2de80266c66a7c2a3858/docs/pages/_document.js#L119-L135)_):
```js
  // Render app and page and get the context of the page with collected side effects.
  const sheets = new ServerStyleSheets();
  const originalRenderPage = ctx.renderPage;

  ctx.renderPage = () =>
    originalRenderPage({
      enhanceApp: (App) => (props) => sheets.collect(<App {...props} />),
    });

  const initialProps = await Document.getInitialProps(ctx);

  let css = sheets.toString();
  // It might be undefined, e.g. after an error.
  if (css && process.env.NODE_ENV === 'production') {
    const result1 = await prefixer.process(css, { from: undefined });
    css = result1.css;
    css = cleanCSS.minify(css).styles;
  }
```

As the next.js buildin postcss configuration is very nice, I was wondering if I can use this configuration in `_document.js` somehow to minify the css from JSS? Or do you have any other ideas on performance improvement and/or minifying and autoprefixing server rendered css?

**Comment by pcjmfranken** [comment, marked as answer]
Faced the same challenge and found a solution. Probably too late for you but in case its useful for anyone else down the line:

1. Rewrite your `postcss.config.js` in `interoperable object-based format` as explained in the Next docs (bottom of [this page](https://nextjs.org/docs/advanced-features/customizing-postcss-config#customizing-plugins) like so:

```js
module.exports = {
  plugins: {
    tailwindcss: {},
    'postcss-flexbugs-fixes': {},
    'postcss-preset-env': {
      autoprefixer: {
        flexbox: 'no-2009',
      },
      stage: 3,
      features: {
        'custom-properties': false,
      },
    },
  },
}
```

2. Install all your PostCss plugins as regular dependencies
3. In your custom _document import your PostCss, your PostCss config, and the plugins:

```tsx
import postcss from 'postcss'
import postcssFlexbugsFixes from 'postcss-flexbugs-fixes'
import postcssPresetEnv from 'postcss-preset-env'
import postcssTailwindcss from 'tailwindcss'

import postcssConfig from '../../postcss.config'
```

4. Now you can use the PostCss JS API and use the same PostCss plugin config you use for your NextJS like so:
```tsx
const postcssPluginConfig = postcss([
  postcssTailwindcss(postcssConfig.plugins.tailwindcss),
  postcssFlexbugsFixes(postcssConfig.plugins['postcss-flexbugs-fixes']),
  postcssPresetEnv(postcssConfig.plugins['postcss-preset-env']),
])

const processCss = async (style: string): Promise<string> => {
  const processedStyles = await postcssPluginConfig.process(style)
  return processedStyles.css
}
```

I'd be happy to provide more examples or further explanation if desired.

PS Untested with PostCss 8.

**Reply by ties-v** [reply]
Thanks for your explanation! Something like this seems the way to go. 

Unfortunately it is not (easily) possible to fetch the default next.js postcss config so you don't have to maintain an extra config file if you need to. The default config is provided in a non exported function: https://github.com/vercel/next.js/blob/b6df8104068fb43fc07183390d661470a3bccb95/packages/next/build/webpack/config/blocks/css/plugins.ts#L74-L105

Maybe this use case will be improved in the future ...
